src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -367,135 +376,383 @@ public class SqlMutationEngine : IMutationEngine
 
-                // All the exceptions that can be thrown by .Complete() and .Dispose() methods of transactionScope
-                // derive from TransactionException. Hence, TransactionException acts as a catch-all.
-                // When an exception related to Transactions is encountered, the mutation is deemed unsuccessful and
-                // a DataApiBuilderException is thrown
-                catch (TransactionException)
+                if (context.OperationType is EntityActionOperation.Upsert || context.OperationType is EntityActionOperation.UpsertIncremental)
                 {
-                    throw _dabExceptionWithTransactionErrorMessage;
-                }
+                    DbResultSet? upsertOperationResult;
+                    DbResultSetRow upsertOperationResultSetRow;
+                    JsonDocument? selectOperationResponse = null;
 
-                DbResultSetRow? dbResultSetRow = upsertOperationResult is not null ?
-                    (upsertOperationResult.Rows.FirstOrDefault() ?? new()) : null;
+                    // This variable indicates whether the upsert resulted in an update operation. If true, then the upsert resulted in an update operation.
+                    // If false, the upsert resulted in an insert operation.
+                    bool hasPerformedUpdate = false;
 
-                if (upsertOperationResult is not null &&
-                    dbResultSetRow is not null && dbResultSetRow.Columns.Count > 0)
-                {
-                    Dictionary<string, object?> resultRow = dbResultSetRow.Columns;
+                    try
+                    {
+                        // Creating an implicit transaction
+                        using (TransactionScope transactionScope = ConstructTransactionScopeBasedOnDbType(sqlMetadataProvider))
+                        {
+                            upsertOperationResult = await PerformUpsertOperation(
+                                                                parameters: parameters,
+                                                                context: context,
+                                                                sqlMetadataProvider: sqlMetadataProvider);
+
+                            if (upsertOperationResult is null)
+                            {
+                                // Ideally this case should not happen, however may occur due to unexpected reasons,
+                                // like the DbDataReader being null. We throw an exception
+                                // which will be returned as an Unexpected InternalServerError
+                                throw new DataApiBuilderException(
+                                    message: "An unexpected error occurred while trying to execute the query.",
+                                    statusCode: HttpStatusCode.InternalServerError,
+                                    subStatusCode: DataApiBuilderException.SubStatusCodes.UnexpectedError);
```

Comment:
Should the substatus code here be UnexpectedRestUpsertOperationFailure?

Reply:
Good catch. 

I have created 4 new status codes - UnexpectedRestUpsertOperationFailure, UnexpectedRestUpsertIncrementalOperationFailure, UnexpectedRestUpdateOperationFailure, UnexpectedRestUpdateIncrementalOperationFailure to be specific on - 1) PUT/PATCH reqeust 2) presence/absence of If-Match header.

Reply:
Tagging @seantleonard - since he already made a similar comment here https://github.com/Azure/data-api-builder/pull/1731#discussion_r1343066142

Reply:
@ayush3797 the internal server error is represented by the status code. My suggestion was that we could be more specific about why the server error occurred using the sub-status code.
After further consideration, I agree with your point about "unexpected" not being appropriate for the sub-status code names. 
The substatus code can help the customer and their end users better troubleshoot the issue to determine whether their dab runtime config needs adjustment to accommodate db object design

Reply:
Consider the scenario where the client application makes different types of REST requests (as needed for its business use-case). When there are failures with some requests, it would be later analyzed by an engineer. At that point, the engineer might only have info related to the errors in the logs. Being specific about the type of the request in the errors would be beneficial for investigating.

Another area where it could be helpful is when we are trying to collect/analyze telemetry at a granular level. For example, we want to collect the count of PATCH requests that failed with unexpected errors.

Reply:
Discussed this during the scrum meeting on 12th Oct, 2023 - Right now, a generic `UnxpectedError` sub status code is to be returned due to the following reasons. 

1. Since, this is an `InternalServerStatus` error, there is nothing actionable for the end users.
2. For DAB folks to investigate, the stack trace should contain sufficient information to investigate.

If there is a need for specific sub status codes, it can be added in the future.